controller/two_lane_queue.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"k8s.io/client-go/util/workqueue"
+)
+
+// item wraps workqueue.Interface.Get() return
+type item struct {
+	i  interface{}
+	sd bool
+}
+
+// twoLaneQueue is a rate limited queue that wraps around two queues
+// -- fast queue, whose contents are processed with priority.
+// -- slow queue, whose contents are processed if fast queue has no items.
+// All the default methods operate on the fast queue, unless noted otherwise.
+type twoLaneQueue struct {
+	workqueue.RateLimitingInterface
+	slowLane workqueue.RateLimitingInterface
+	// consumerQueue is necessary to ensure that we're not reconciling
+	// the same object at the exact same time (e.g. if it had been enqueued
+	// in both fast and slow and is the only object there).
+	consumerQueue workqueue.Interface
+
+	name string
+
+	fastChan chan item
+	slowChan chan item
+}
+
+// Creates a new twoLaneQueue.
+func newTwoLaneWorkQueue(name string) *twoLaneQueue {
+	rl := workqueue.DefaultControllerRateLimiter()
+	tlq := &twoLaneQueue{
+		RateLimitingInterface: workqueue.NewNamedRateLimitingQueue(
+			rl,
+			name+"-fast",
+		),
+		slowLane: workqueue.NewNamedRateLimitingQueue(
+			rl,
+			name+"-slow",
+		),
+		consumerQueue: workqueue.NewNamed(name + "-consumer"),
+		name:          name,
+		fastChan:      make(chan item),
+		slowChan:      make(chan item),
+	}
+	tlq.runConsumer()
+	tlq.runProducers()
+	return tlq
+}
+
+func process(q workqueue.Interface, ch chan item) {
+	go func(q workqueue.Interface) {
+		for {
+			i, d := q.Get()
+			// If the queue is empty and we're shutting down — stop the loop.
+			if d {
+				break
+			}
+			ch <- item{i: i, sd: d}
+		}
+		// Sender closes the channel
+		close(ch)
+	}(q)
+}
+
+func (tlq *twoLaneQueue) runProducers() {
+	process(tlq.RateLimitingInterface, tlq.fastChan)
+	process(tlq.slowLane, tlq.slowChan)
+}
+
+func (tlq *twoLaneQueue) runConsumer() {
+	tlq.consumerQueue = workqueue.NewNamed(tlq.name + "consumer")
+	// Shutdown flags.
+	fast, slow := true, true
+	go func() {
+		for fast || slow {
+			// By default drain the fast lane.
+			if fast {
+				select {
+				case item, ok := <-tlq.fastChan:
+					if !ok {
+						fast = false
+						continue
+					}
+					tlq.consumerQueue.Add(item.i)
+					continue
+				default:
```

Comment:
It's probably worth documenting the hell out of the flow of control here for folks that don't understand all of the magic around Go channels.